onnxruntime/core/framework/session_state.cc

```diff
@@ -616,6 +691,14 @@ Status SessionState::GeneratePatternGroupCache(gsl::span<const OrtValue> tensor_
       const auto* ml_type = exe_plan->allocation_plan[ml_value_idx].value_type;
       if (!ml_type->IsTensorType())
         continue;
+
+      auto* ep = GetExecutionProviders().Get(*node);
+      auto device_id = exe_plan->allocation_plan[ml_value_idx].location.device.Id();
+      auto ep_main_allocator = ep->GetAllocator(device_id, OrtMemType::OrtMemTypeDefault);
+      ORT_ENFORCE(ep_main_allocator);
```

Comment:
sure, I am actually a little confused by our allocator management, i thought EP owns the allocator, but seems not, the Session's "RegisterAllocators" call seems to be the owner of all the allocators. should i check the allocators from session_state instead of from EP?

Reply:
Allocator management in general still needs clarification. Ideally you want one allocator per device +memory type otherwise you get duplicate arenas and wasted memory usage. There's some allocator sharing between EPs now to try and address that (EP can optionally implement IExecutionProvider::RegisterAllocator) to some degree. 

The execution frame gets the allocator from SessionState::GetAllocator when allocating node outputs so matching that might be more consistent.

Reply:
yes i think we should enhance our "Device" abstraction to make it clear. 

for the code here, it actually just wants to skip the case for pinned memory, i simplify it by just check the memory type.